File: backend/app/schemas/multi_chat.py
"""Schemas for multi-paper chat (group/selection-based AI conversations)."""

import base64
import secrets
from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


def generate_multi_session_name() -> str:
  """Generate a random base64 session name."""
  return base64.urlsafe_b64encode(secrets.token_bytes(6)).decode()


# --- Paper summary (lightweight representation) ---


class PaperSummary(BaseModel):
  id: int
  title: str
  has_file: bool = False

  class Config:
    from_attributes = True


# --- Messages ---


class MultiChatMessageBase(BaseModel):
  role: str  # 'user' or 'assistant'
  content: str
  references: Optional[Dict[str, Any]] = {}


class MultiChatMessageCreate(BaseModel):
  content: str
  references: Optional[Dict[str, Any]] = {}


class MultiChatMessage(MultiChatMessageBase):
  id: int
  session_id: int
  parent_message_id: Optional[int] = None
  thread_count: int = 0
  created_at: datetime

  model_config = {"from_attributes": True}


# --- Sessions ---


class MultiChatSessionCreate(BaseModel):
  paper_ids: Optional[List[int]] = None
  group_id: Optional[int] = None
  name: str = Field(default_factory=generate_multi_session_name)


class MultiChatSessionUpdate(BaseModel):
  name: str


class MultiChatSession(BaseModel):
  id: int
  name: str
  group_id: Optional[int] = None
  paper_ids: List[int] = []
  papers: List[PaperSummary] = []
  messages: List[MultiChatMessage] = []
  created_at: datetime
  updated_at: datetime

  model_config = {"from_attributes": True}


# --- Request / Response ---


class MultiChatRequest(BaseModel):
  message: str
  references: Optional[Dict[str, Any]] = {}
  session_id: Optional[int] = None
  paper_ids: Optional[List[int]] = None
  group_id: Optional[int] = None


class MultiChatResponse(BaseModel):
  message: MultiChatMessage
  session: MultiChatSession

File: frontend-v2/src/lib/api/multi-chat.ts
import { api } from './client';
import type { ChatReferences } from './chat';

export interface PaperSummary {
  id: number;
  title: string;
  has_file: boolean;
}

export interface MultiChatMessage {
  id: number;
  session_id: number;
  role: 'user' | 'assistant';
  content: string;
  references?: ChatReferences;
  created_at: string;
  parent_message_id: number | null;
  thread_count: number;
}

export interface MultiChatSession {
  id: number;
  name: string;
  group_id: number | null;
  paper_ids: number[];
  papers: PaperSummary[];
  messages: MultiChatMessage[];
  created_at: string;
  updated_at: string;
}

export interface MultiChatRequest {
  message: string;
  references?: ChatReferences;
  session_id?: number;
  paper_ids?: number[];
  group_id?: number;
}

export const multiChatApi = {
  getGroupHistory: (groupId: number): Promise<MultiChatSession | null> =>
    api.get<MultiChatSession | null>(`/groups/${groupId}/chat`),

  getGroupSessions: (groupId: number): Promise<MultiChatSession[]> =>
    api.get<MultiChatSession[]>(`/groups/${groupId}/multi-sessions`),

  createGroupSession: (groupId: number, name?: string, paperIds?: number[]): Promise<MultiChatSession> =>
    api.post<MultiChatSession>(`/groups/${groupId}/multi-sessions`, {
      name: name ?? undefined,
      paper_ids: paperIds ?? undefined,
      group_id: groupId,
    }),

  getSession: (sessionId: number): Promise<MultiChatSession> =>
    api.get<MultiChatSession>(`/multi-chat/sessions/${sessionId}`),

  updateSession: (sessionId: number, name: string): Promise<MultiChatSession> =>
    api.patch<MultiChatSession>(`/multi-chat/sessions/${sessionId}`, { name }),

  deleteSession: async (sessionId: number): Promise<void> => {
    await api.delete(`/multi-chat/sessions/${sessionId}`);
  },

  clearSessionMessages: async (sessionId: number): Promise<void> => {
    await api.delete(`/multi-chat/sessions/${sessionId}/messages`);
  },
};
